src/frontend/Semantic_error.ml

```diff
@@ -152,6 +162,96 @@ module TypeError = struct
           (String.concat ~sep:"" lines)
           Fmt.(list UnsizedType.pp ~sep:comma)
           arg_tys
+    | IllTypedVariadicODE (name, arg_tys, args) ->
+        let arg_types = List.map ~f:(fun (_, t) -> t) args in
+        let generate_reduce_sum_sig =
+          List.concat
+            [ [ UnsizedType.UFun
+                  ( (AutoDiffable, UReal) :: (AutoDiffable, UVector) :: args
+                  , ReturnType UVector ) ]
+            ; [UVector; UReal; UArray UReal]
+            ; arg_types ]
+        in
+        Fmt.pf ppf
+          "Ill-typed arguments supplied to function '%s'. Expected \
+           arguments:@[<h>%a@]\n\
+           @[<h>Instead supplied arguments of incompatible type: %a@]"
+          name
+          Fmt.(list UnsizedType.pp ~sep:comma)
+          generate_reduce_sum_sig
+          Fmt.(list UnsizedType.pp ~sep:comma)
+          arg_tys
+    | IllTypedVariadicODEGeneric (name, arg_tys) ->
+        let variadic_ode_generic_signature =
+          Fmt.strf "(%a, %a, ...) => %a, %a, %a, %a, ...\n"
+            Pretty_printing.pp_unsizedtype UReal
+            (* fun: time *)
+            Pretty_printing.pp_unsizedtype UVector
+            (* fun: state *)
+            Pretty_printing.pp_unsizedtype UVector
+            (* fun: return *)
+            Pretty_printing.pp_unsizedtype UVector
+            (* initial_state *)
+            Pretty_printing.pp_unsizedtype UReal
+            (* initial_time *)
+            Pretty_printing.pp_unsizedtype (UArray UReal)
+          (* times *)
+        in
+        Fmt.pf ppf
+          "Ill-typed arguments supplied to function '%s'. Available arguments:\n\
+           %sWhere T1 and T2 are of any type.@[<h>Instead supplied arguments \
+           of incompatible type: %a@]"
+          name variadic_ode_generic_signature
+          Fmt.(list UnsizedType.pp ~sep:comma)
+          arg_tys
+    | IllTypedVariadicODETol (name, arg_tys, args) ->
+        let arg_types = List.map ~f:(fun (_, t) -> t) args in
+        let generate_ode_sig =
+          List.concat
+            [ [ UnsizedType.UFun
+                  ( (AutoDiffable, UReal) :: (AutoDiffable, UVector) :: args
+                  , ReturnType UVector ) ]
+            ; [UVector; UReal; UArray UReal; UReal; UReal; UInt]
+            ; arg_types ]
+        in
+        Fmt.pf ppf
+          "Ill-typed arguments supplied to function '%s'. Expected \
+           arguments:@[<h>%a@]\n\
+           @[<h>Instead supplied arguments of incompatible type: %a@]"
+          name
+          Fmt.(list UnsizedType.pp ~sep:comma)
+          generate_ode_sig
+          Fmt.(list UnsizedType.pp ~sep:comma)
+          arg_tys
+    | IllTypedVariadicODEGenericTol (name, arg_tys) ->
```

Comment:
Again I'm wondering if there's a less redundant way to write this, since it only differs from IllTypedVariadicODEGeneric by the list of types - but it might not be worth it here, up to you

Reply:
Will try to reduce the redundancy. Good call